rir/src/compiler/util/cfg.cpp

```diff
@@ -49,49 +49,303 @@ bool CFG::isPredecessor(BB* a, BB* b) const {
+    //     clause of the Dominator Theorem.
+    //
+    //   Step 4.
+    //     Explicitly define the immediate dominator by applying the second
+    //     clause of the Dominator Theorem. Do this while iterating over nodes
+    //     in increasing dfnum.
+    //
+    //     Dominator Theorem.
+    //       On the DFS tree path below semi(n) and above or including `n`, let
+    //       `y` be the node the the smallest numbered semidominator, i.e. with
+    //       minimum dfnum[semi(y)]. Then:
+    //         1. idom(n) = semi(n)    if semi(y) == semi(n)
+    //         2. idom(n) = idom(y)    if semi(y) != semi(n)
+    //
+    //
+    // As the algorithm runs, it maintains a spanning forest of nodes -- at the
+    // end, the forest will be fully connected as a single spanning tree. The
+    // algorithm is designed so that when a node `n` is being processed, only
+    // nodes with a higher dfnum than `n`, i.e. nonancestors of `n`, will be in
+    // the forest.
+    //
+    // This spanning forest is represented by the `ancestor` array, updated by
+    // the `link` function, and queried by the `ancestorWithLowestSemi`
+    // function.
+    //
+    // Note that the naive implementation of `ancestorWithLowestSemi` is O(N),
+    // but this implementation performs "path compression," so its amortized
+    // complexity is O(log N). There is an even more sophisticated version with
+    // better algorithmic complexity [LT79], that performs "balanced path
+    // compression," but in practice, it is actually slower [GTW06].
+    //
+    // As `ancestorWithLowestSemi` searches the spanning forest, it updates
+    // the `ancestor` array to compress the path. However, as it does so, it
+    // must remember the best result in the path that was skipped, i.e. the
+    // node with semidominator with lowest dfnum.
+    //
+    //
+    // References:
+    //
+    // [AP04] Appel, A. and Palsberg, J. (2004). "Efficient Computation of the
+    // Dominator Tree." In "Modern Compiler Implementation in Java," 2nd ed.
+    //
+    // [GTW06] Georgiadis, L., Tarjan, R. E., and Werneck R. F. (2006).
+    // "Finding dominators in practice." J. Graph Algorithms Appl.
+    // http://doi.org/10.7155/jgaa.00119
+    //
+    // [LT79] Lengauer, T. and Tarjan, R. E. (1979). "A fast algorithm for
+    // finding dominators in a flowgraph." ACM Trans. Program. Lang. Syst.
+    // https://doi.org/10.1145/357062.357071
+
+    /***************************************************************************
+     * Declarations and initializations.
+     **************************************************************************/
+
+    // Note that this is an overapproximation of the actual number of BBs,
+    // because some BBs may have been deleted.
+    int size = start->nextBBId;
+
+    // Counter for numbering BBs, by depth-first search order.
+    int N = 0;
+
+    // Indexed by BB id. `dfnum[n->id]` is the dfnum of node `n`.
+    std::vector<int> dfnum(size);
```

Comment:
@mhyee why do you keep a number on the side, when you can use the bb->id to store it? The compiler is designed such that it is always allowed to renumber the bbs between passes, the only invariant is that the max is is smaller than nextBBId. The id is mostly used to speed up lookups in analysis and visitors (by using the id as an array index). Therefore the only caveat is to not use the default visitor while you renumber (see BBTransform::renumber for an example).

Reply:
Honestly, I just didn't think of this. Doing it on-the-fly means I need some other way of tracking which BBs have already been seen, so the code might be slightly messier.

But you're right -- if some other analysis requires dominance, it might break if the BBs are renumbered...

Reply:
> But you're right -- if some other analysis requires dominance, it might break if the BBs are renumbered...

ah, i did not realize that you are already using the id permanently. I think that is an ok assumption, that id's are mostly stable. but it would be nice if your analysis could somehow rely on the fact that they are already correctly ordered. maybe we could even have this as an invariant?

Reply:
But there's quite a few places where we insert or merge blocks, and that would ruin the depth-first order. So I don't think we can guarantee that we always have depth-first order, unless we do a renumber after every BB transformation.

Reply:
right. just throwing around ideas...